cpp/src/decisiontree/decisiontree.cuh

```diff
@@ -0,0 +1,425 @@
+      int node_id = q_node.unique_node_id;
+
+      if (!is_leaf_node) {
+        tl_tree.AddChilds(node_id);
+
+        // Push left child to next_level queue.
+        next_level_queue[next_end] =
+          Node_ID_info<T, L>(rf_tree.sparsetree[q_node.node->left_child_id],
+                             tl_tree.LeftChild(node_id));
+        ++next_end;
+
+        // Push right child to next_level queue.
+        next_level_queue[next_end] =
+          Node_ID_info<T, L>(rf_tree.sparsetree[q_node.node->left_child_id + 1],
+                             tl_tree.RightChild(node_id));
+        ++next_end;
+
+        // Set node from current level as numerical node. Children IDs known.
+        tl_tree.SetNumericalSplit(node_id, q_node.node->colid,
+                                  q_node.node->quesval, true,
+                                  tl::Operator::kLE);
+
+      } else {
+        if (num_class == 1) {
+          tl_tree.SetLeaf(node_id, static_cast<T>(q_node.node->prediction));
+        } else {
+          std::vector<T> leaf_vector(num_class, 0);
+          leaf_vector[q_node.node->prediction] = 1;
+          tl_tree.SetLeafVector(node_id, leaf_vector);
+        }
+      }
+    }
+
+    cur_level_queue.swap(next_level_queue);
+    cur_front = next_front;
+    cur_end = next_end;
+    next_front = 0;
+    next_end = 0;
+  }
+  return tl_tree;
+}
+
+struct DataInfo {
+  unsigned int NLocalrows;
+  unsigned int NGlobalrows;
+  unsigned int Ncols;
+};
+
+template <class T, class L>
+class DecisionTree {
+ protected:
+  DataInfo dinfo;
+  int depth_counter = 0;
+  int leaf_counter = 0;
+  int n_unique_labels = -1;  // number of unique labels in dataset
+  double prepare_time = 0;
+  double train_time = 0;
+  MLCommon::TimerCPU prepare_fit_timer;
+  DecisionTreeParams tree_params;
+
+ public:
+  /**
+     * @brief Fits a DecisionTree on given input data and labels
+     * @param[in] handle             cuML handle
+     * @param[in] data               pointer to input training data
+     * @param[in] ncols              number of features (columns)
+     * @param[in] nrows              number of samples (rows)
+     * @param[in] labels             pointer to label data
+     * @param[in] rowids             pointer to array of row indices mapping to data
+     * @param[in] n_sampled_rows     count of rows sampled
+     * @param[in] unique_labels      count of unique labels
+     * @param[in] is_classifier      true if task is classification, else false
+     * @param[in,out] tree           pointer to tree structure
+     * @param[in] tree_parameters    structure of tree parameters
+     * @param[in] seed               random seed
+     * @param[in] d_global_quantiles device pointer to global quantiles
+     */
+  void fit(const raft::handle_t &handle, const T *data, const int ncols,
+           const int nrows, const L *labels, unsigned int *rowids,
+           const int n_sampled_rows, int unique_labels, bool is_classifier,
```

Comment:
I'd prefer to avoid the `is_classifier` flag. Is the default criterion selected in the python code? I'm not sure we need to worry about that in C++ code.

Reply:
if an unsupported criterion is set in the python code, a `CRITERION::CRITERION_END` is passed below from [here](https://github.com/rapidsai/cuml/blob/76e6b51fa16497b83e2b4ed708f2b51d445e6dad/python/cuml/ensemble/randomforest_common.pyx#L116) and is resolved here. I can do resolution in the python layer itself, one for each regressor and classifier.

Reply:
Personally I think it should be resolved in python, and I wouldn't mind changing the interface to use strings instead of integers to copy skl. That can happen another day though.